### PR TITLE
Tests: Contract test for CCoinsView and CCoinsViewBacked

### PR DESCRIPTION
**Pull request by mmachicao**
Added some admittedly trivial tests to protect against unintended and breaking modifications of the respective contracts.

Remark:
AFAIK the only specialization in production code of CCoinsView is CCoinsViewDB
On the other hand, there are several parts in the production code where a DUMMY instance of CCoinsView is instantiated only to comply with the signature of the class CCoinsViewCache.

IMHO this smells like overengineering (aka: the future never comes) or forced code recycling and could potentially lead to unintended behavior along the road.

Will invest some time to take a deeper look into this.

e.g.

```
./bitcoin/src/rpc/rawtransaction.cpp:    CCoinsView viewDummy;
./bitcoin/src/rpc/rawtransaction.cpp:    CCoinsViewCache view(&viewDummy);

./bitcoin/src/bitcoin-tx.cpp:    CCoinsView viewDummy;
./bitcoin/src/bitcoin-tx.cpp:    CCoinsViewCache view(&viewDummy);

./bitcoin/src/validation.cpp:        CCoinsView dummy;
./bitcoin/src/validation.cpp:        CCoinsViewCache view(&dummy);

./bitcoin/src/bench/ccoins_caching.cpp:    CCoinsView coinsDummy;
./bitcoin/src/bench/ccoins_caching.cpp:    CCoinsViewCache coins(&coinsDummy);
```

### CODE REVIEW

File: src/test/coins_tests.cpp [hunk 1]
```diff
@@ -98,6 +98,78 @@ class CCoinsViewCacheTest : public CCoinsViewCache
...
+    uint256 hashBlock;
+    BOOST_CHECK_EQUAL(vBacked.BatchWrite(map, hashBlock), false);
+    BOOST_CHECK(vBacked.Cursor() == nullptr);
+    BOOST_CHECK_EQUAL(vBacked.EstimateSize(), 0);
+
+    // Provide an alternaitve implementation of CCoinsView with the opposite behavior
+    // and test that CCoinsViewBackend behaves accordingly
+    class Opposite : public CCoinsView
+    {
+        bool GetCoin(const COutPoint& outpoint, Coin& coin) const override { return true; }
```
Comment (practicalswift) [inline src/test/coins_tests.cpp]: Choose another name for `coin` to avoid shadowing `coin` on L127?

File: src/test/coins_tests.cpp [hunk 2]
```diff
@@ -98,6 +98,78 @@ class CCoinsViewCacheTest : public CCoinsViewCache
...
+        bool BatchWrite(CCoinsMap& mapCoins, const uint256& hashBlock) override { return true; }
```
Comment (practicalswift) [inline src/test/coins_tests.cpp]: Same here but for `hashBlock`.

File: src/test/coins_tests.cpp [hunk 3]
```diff
@@ -98,6 +98,78 @@ class CCoinsViewCacheTest : public CCoinsViewCache
...
+    BOOST_CHECK_EQUAL(vBacked.GetBestBlock(), uint256());
+    BOOST_CHECK_EQUAL(vBacked.GetHeadBlocks().size(), 0);
+
+    CCoinsMap map;
+    uint256 hashBlock;
+    BOOST_CHECK_EQUAL(vBacked.BatchWrite(map, hashBlock), false);
+    BOOST_CHECK(vBacked.Cursor() == nullptr);
+    BOOST_CHECK_EQUAL(vBacked.EstimateSize(), 0);
+
+    // Provide an alternaitve implementation of CCoinsView with the opposite behavior
```
Comment (practicalswift) [inline src/test/coins_tests.cpp]: Should be "alternative" :-)

Comment (mmachicao) [reply src/test/coins_tests.cpp]: Done. Unfortunately, the build environment broke down.

**Pull request closed without merging**